src/impl_/pymodule.rs

```diff
@@ -82,6 +82,10 @@ impl ModuleDef {
         (self.initializer.0)(py, module.as_ref(py))?;
         Ok(module)
     }
+
+    pub fn add_to_module(&'static self, module: &PyModule) -> PyResult<()> {
+        module.add_object(self.make_module(module.py())?.into())
+    }
```

Comment:
I think to avoid the reinitialization we could use a hypothetical `self.find_or_make_module()` here which uses `PyState_FindModule` to retrieve the existing module object (if it exists).

I think this exists on all CPython versions, for PyPy it only exists on PyPy 3.9 (with the wrong name) and PyPy 3.10. Probably it'll be good enough to just reinitialize on PyPy 3.8 and older, they're out of support anyway.

I'll send a PR now to fix the PyPy definitions for `PyState_FindModule`.